File: src/qc.py
import unittest
from math import pi, sin, cos, sqrt, atan2

import matplotlib.pyplot as plt
import numpy as np
from numpy import complex, arcsin, log2, floor, vdot


def padded_bin(n, k):
    return bin(k)[2:].zfill(n)


def show_graph(amplitudes, title, title_x, title_y):
    y_position = np.arange(len(amplitudes))
    plt.bar(y_position, [abs(v) for v in amplitudes.values()], align='center',
            color=[[x/255 for x in complex_to_rgb(v)] for v in amplitudes.values()])
    plt.xticks(y_position, amplitudes.keys())
    plt.xticks(rotation=90)
    plt.xlabel(title_x)
    plt.ylabel(title_y)
    plt.title(title)
    plt.tight_layout()
    plt.show()


def plot_state(state, name = ""):
    n = int(log2(len(state)))
    s = dict([(padded_bin(n, k)[::-1], state[k]) for k in range(len(state))])
    d = {k + '=' + str(int(k, 2)): # [::-1]
             complex(round(v.real, 5),round(v.imag, 5)) for k, v in s.items()}
    d = dict(sorted(d.items(), key=lambda kv: (kv[0], kv[1])))
    print(name, "amplitudes:", d)
    show_graph(d, "Quantum Simulation" + " " + name, "Outcomes", "Amplitudes")
    p = {k: round(abs(v) ** 2, 5) for k, v in d.items()}
    print(name, "probabilities:", p)
    show_graph(p, "Quantum Simulation" + " " + name, "Outcomes", "Probabilities")


def pixel_state(s, cols = 1):
    n = int(log2(len(s)))
    rows = int(len(s) / cols)
    s = {int(padded_bin(n, k)[::-1], 2):
             s[k] for k in range(len(s))}

    from PIL import Image, ImageOps
    # PIL accesses images in Cartesian co-ordinates, so it is Image[columns, rows]
    img = Image.new('RGB', (cols, rows), "black")  # create a new black image
    pixels = img.load()  # create the pixel map
    for i in range(img.size[0]):  # for every col:
        for j in range(img.size[1]):  # For every row
            rgb = [int(x) for x in complex_to_rgb(s[i*rows + j], True)]
            pixels[i, j] = (rgb[0], rgb[1], rgb[2])  # set the colour accordingly
    img = ImageOps.flip(img)
    # img.show()
    img.save("state.png")


def complex_to_rgb(c, scaled_saturation = False):
    a = c.real
    b = c.imag

    val = 100;

    hue = atan2(b, a) * 180 / pi;
    if hue < 0: hue += 360

    sat = 100
    if scaled_saturation:
        sat = sqrt(a ** 2 + b ** 2) * 100

    return hsv_to_rgb(hue, sat, val)


# https://gist.github.com/eyecatchup/9536706 Colors
def hsv_to_rgb(h, s, v):
    # Make sure our arguments stay in-range
    h = max(0, min(360, h))
    s = max(0, min(100, s))
    v = max(0, min(100, v))
    
    # We accept saturation and value arguments from 0 to 100 because that's
    # how Photoshop represents those values. Internally, however, the
    # saturation and value are calculated from a range of 0 to 1.
    # We make that conversion here.
    s /= 100
    v /= 100
    
    if s == 0:
        # Achromatic (grey)
        r = g = b = v
        return [
            round(r * 255),
            round(g * 255),
            round(b * 255)
        ]

    h /= 60 #sector 0 to 5
    i = floor(h)
    f = h - i #factorial part of h
    p = v * (1 - s)
    q = v * (1 - s * f)
    t = v * (1 - s * (1 - f))

    if i == 0:
        r = v
        g = t
        b = p
    elif i == 1:
        r = q
        g = v
        b = p
    elif i == 2:
        r = p
        g = v
        b = t
    elif i == 3:
        r = p
        g = q
        b = v
    elif i  == 4:
        r = t
        g = p
        b = v
    else: #case 5:
        r = v
        g = p
        b = q

    return [
        round(r * 255),
        round(g * 255),
        round(b * 255)
    ]


def init_state(n):
    state = [0 for k in range(2 ** n)]
    state[0] = 1
    return state


def transform(state, t, gate, cond = lambda m: True):
    n = int(log2(len(state)))
    distance = int(2 ** t)
    suffix_count = int(2 ** t)
    prefix_count = int(2 ** (n - t - 1))

    for p in range(prefix_count):
        for s in range(suffix_count):
            m0 = p * suffix_count*2 + s
            if cond(m0):
                # the 1 side of the pair
                m1 = m0 + distance
                x = state[m0]
                # print(b, "<-->", format(m1, '#0' + str(n + 2) + 'b')[2:][::-1])
                y = state[m1]
                # new amplitudes
                state[m0] = x * gate[0][0] + y * gate[0][1]
                state[m1] = x * gate[1][0] + y * gate[1][1]


def c_transform(state, c, t, gate):
    # def cond(n, m):
    #     # b = format(m, '#0' + str(n + 2) + 'b')[2:][::-1]
    #     # return b[c] == '1'
    #     return m >> c & 1
    #
    # transform(state, t, gate, cond)
    mc_transform(state, {c:1}, t, gate)

# cs: dictionary with map of qubits to be 1 or 0
def mc_transform(state, cs, t, gate):
    def cond(m):
        ret = True
        for k, v in cs.items():
            bit_set = m >> k & 1
            ret = ret and (not bit_set ^ v)
        return ret

    transform(state, t, gate, cond)


def m_transform(state, targets, gate):
    for j in targets:
        transform(state, j, gate)


def qft(state, targets):
    for j in targets:
        transform(state, j, h)
        for k in targets[:j]:
            c_transform(state, j, k, phase(pi / 2 ** (k - j)))


def iqft1(state, targets):
    l = len(targets)
    for j in targets:
        j = l - 1 - j
        transform(state, j, h) # ry(-pi/2)
        for k in targets[:j]:
            c_transform(state, j, j - 1 - k, phase(-pi / 2 ** (k + 1)))


def iqft(state, targets):
    for j in targets[::-1]:
        transform(state, j, h)
        for k in targets[:j][::-1]:
            c_transform(state, j, k, phase(-pi / 2 ** (j - k)))


# gates
h = [[1/sqrt(2), 1/sqrt(2)], [1/sqrt(2), -1/sqrt(2)]]
x = [[0, 1], [1, 0]]
z = [[1, 0], [0, -1]]


def rx(theta):
    return [[cos(theta/2), complex(0, -sin(theta/2))], [complex(0, -sin(theta/2)), cos(theta/2)]]


def ry(theta):
    return [[cos(theta/2), -sin(theta/2)], [sin(theta/2), cos(theta/2)]]


def phase(theta):
    return [[1, 0], [0, complex(cos(theta), sin(theta))]]


def param_encoding(state, targets, v):
    # theta = v * 2 * pi / 2**len(targets)
    #
    # for j in targets:
    #     transform(state, j, h)
    #
    # for i in targets:
    #     transform(state, i, phase(2 ** i * theta))
    geom_sim(state, v)

    # iqft(state, targets)
    ift_sim(state)


def fibonacci(state, targets):
    theta = 2*arcsin((sqrt(5) - 1)/2)

    n = len(targets)
    for i in range(n):
        transform(state, targets[i], ry(theta))

    for i in range(n-1):
        c_transform(state, targets[i], targets[i+1], ry(-theta))


def geom_quantum(state, v):
    targets = range(int(log2(len(state))))
    theta = v * 2 * pi / 2**len(targets)

    m_transform(state, targets, h)
    for i in targets:
        transform(state, i, phase(2 ** i * theta))


def geom_sim(state, v):
    N = len(state)
    param = cos(v*2*pi/N) + 1j*sin(v*2*pi/N)
    for k in range(N):
        state[k] = param**(k)


def ift_sim(state):
    N = len(state)
    n = int(log2(N))
    s = [state[k] for k in range(N)]
    for i in range(N):
        fourier = np.empty(N, dtype=np.complex_)
        root = cos(i*2*pi/N) + 1j*sin(i*2*pi/N)
        for k in range(N):
            fourier[k] = root**k

        state[int(padded_bin(n, i)[::-1], 2)] = vdot(fourier, s)/N


def grover_sim(state, items, iterations):
    n = int(log2(len(state)))
    for i in range(len(items)):
        items[i] = int(padded_bin(n, items[i])[::-1], 2)
    s = state.copy()

    # Grover iterate
    for _ in range(iterations):
        # oracle (reflection in bad state vector)
        for item in items:
            state[item] = -1 * state[item]

        # inversion (reflection in original state)
        inner = vdot(s, state)
        for k in range(len(state)):
            state[k] = 2*inner*s[k] - state[k]


def encode_value(state, targets, k, v):
    theta = v * 2 * pi / 2 ** len(targets)
    for i in range(len(targets)):
        d = dict([(i, int(k[i])) for i in range(len(k))])
        print(d)
        mc_transform(state, d, targets[i], phase(2 ** i * theta))


def encode_value1(state, targets, c, v):
    theta = v * 2 * pi / 2 ** len(targets)
    for i in range(len(targets)):
        c_transform(state, c, targets[i], phase(2 ** i * theta))


class TestSim(unittest.TestCase):

    def test_grover(self):
        n = 3
        s = init_state(n)
        m_transform(s, range(n), h)
        grover_sim(s, items = [3], iterations = 2)

        plot_state(s)
        pixel_state(s, 1)

    def test_dictionary(self):
        n = 5
        state = init_state(n)

        m_transform(state, [0, 1], h)

        targets = [2, 3, 4]
        m_transform(state, targets, h)

        encode_value(state, targets, '00', 1)
        encode_value(state, targets, '01', 5)
        encode_value(state, targets, '10', 2)
        encode_value(state, targets, '11', 7)

        iqft(state, targets)

        plot_state(state)
        pixel_state(state, 4)

    def test_fibonacci(self):

        n = 4
        s = init_state(n)

        fibonacci(s, range(n))

        plot_state(s)
        pixel_state(s)

    def test_param(self):
        n = 4
        s = init_state(n)
        param_encoding(s, range(n), 5.7)
        plot_state(s)
        pixel_state(s, 1)


if __name__ == "__main__":
    # unittest.main()
    TestSim.test_grover()

File: src/qc.jl
#=
qc:
- Author: contralto
- Date: 2020-12-23
=#

using LinearAlgebra, Parameters, Test

global counter = 0
global folder = "images/"

const Amplitude = Complex

const State = Vector{Amplitude}

struct Gate
    name::String
    weights::Vector{Vector{Amplitude}}
end

struct Transform
    gate::Gate
    control::Int
    target::Int
end

############################### GATES ##########################################

h = Gate("H", [[Complex(1/sqrt(2)), 1/sqrt(2)],
                       [1/sqrt(2), -1/sqrt(2)]])

x = Gate("X", [[0, 1], [1, 0]])

z = Gate("Z", [[1, 0], [0, -1]])

function Rx(theta)
     return Gate("Rx(" * string(round(theta,digits=3)) * ")",
                 [[cos(theta / 2), -im * sin(theta / 2)],
                 [-im * sin(theta / 2), cos(theta / 2)]])
end

function Ry(theta)
     return Gate("Ry(" * string(round(theta,digits=3)) * ")",
                 [[cos(theta / 2), -sin(theta / 2)],
                  [sin(theta / 2), cos(theta / 2)]])
end

function phase(theta)
    return Gate("Phase(" * string(round(theta,digits=3)) * ")",
                [[1, 0], [0, (cos(theta) + im * sin(theta))]])
end

############################### ENCODING ##########################################

# # check cth digit of binary string of x is 1
function isDigitOne(x::Int, c::Int, n::Int)
    str = reverse(string(x, base = 2, pad = n))
    return str[c + 1] == '1'
end

function pair_exchange!(state::State, gate::Gate, m0::Int, m1::Int)
    x = state[m0]
    y = state[m1]
    state[m0] = gate.weights[1][1] * x + gate.weights[1][2] * y
    state[m1] = gate.weights[2][1] * x + gate.weights[2][2] * y
end

function c_transform(state::State, c::Int, t::Int, gate::Gate)
    c = Int(log2(length(state))) - c - 1
    transform!(state, t, gate, m -> isDigitOne(m, c, Int(log2(length(state)))))
end

function iqft(state::State, targets::Array)
    for j in reverse(targets)
        transform!(state, j, h)
        for k in reverse(0:j - 1)
            c_transform(state, j, targets[k + 1], phase(-pi / (2 ^ (j-k))))
        end
    end
end


function value_encoding(state, targets, v)
    theta = v * 2 * pi / (2 ^ length(targets))

    for j in targets
        transform!(state, j, h)
    end

    for i in targets
        transform!(state, i, phase(2 ^ i * theta))
    end

    iqft(state, targets)
end


############################### TRANSFORMATIONS ##########################################

function transform!(state::State, target::Int, gate::Gate,
                    cond::Function = f(x::Int)::Bool = true)
    # sections begin when we reach a new set of "first qubits" (i.e. first in a pair)
    # ex. with distance 2 and state [1, 0, 0, 0, 0, 0, 0, 0], the sections would be
    # [1*, 0*, 0, 0], [0*, 0*, 0, 0] (* signifies the first qubit of each pair)

    target = Int(log2(length(state))) - target - 1

    # number of qubits in each section of the state
    section_len = Int(2 ^ (target + 1))

    # distance between paired qubits
    distance = Int(section_len / 2)

    # applies the gate to each pair
    for section in 0:(length(state) / section_len - 1)
        for pair in 0:(section_len / 2 - 1)
            # index of first qubit in pair
            m0 = Int(section_len * (section) + pair)
            if (cond(m0))
                # index of second qubit in pair
                m1 = m0 + distance
                pair_exchange!(state, gate, m0 + 1, m1 + 1)
            end
        end
    end
    title = "Gate: " * gate.name * ", target: " * string(target)
    save_state(state, title)

    state
end




function transform_with_matrix!(state::State, target::Int, gate::Gate,
                                cond::Function = f(x::Int)::Bool = true)
    n = length(state)
    target = Int(log2(n)) - target - 1
    G = Matrix{ComplexF64}(I, n, n)

    factor = 2 ^ (target + 1)
    shift =  2 ^ (target)

    for prefix in 0:(n / factor - 1)
        for suffix in 0:(2^target - 1)
            m0 = Int(factor * (prefix) + suffix)
            if (cond(m0))
                # + 1 for julia indexing
                m0 += 1
                m1 = m0 + shift
                G[m0, m0] = gate.weights[1][1]
                G[m0, m1] = gate.weights[1][2]
                G[m1, m0] = gate.weights[2][1]
                G[m1, m1] = gate.weights[2][2]
            end
        end
    end

    state .= G * state
    title = "Gate: " * gate.name * ", target: " * string(target)
    save_state(state, title)

    state
end

############################### GRAPHING ##########################################

using Gadfly, Colors

# https://gist.github.com/eyecatchup/9536706 Colors
function complex_to_hsv(z::Complex, change_sat::Bool = true)
    r = real(z)
    i = imag(z)

    val = 1.0

    hue = angle(z) * 180 / pi
    if (hue < 0) hue += 360 end

    if (change_sat) sat = abs(z)
    else sat = 1.0 end

    # Make sure our arguments stay in-range
    hue = max(0, min(360, hue))
    sat = max(0.0, min(1.0, sat))
    val = max(0, min(100, val))

    return (hue, sat, val)
end

function bar_state(state, amp::Bool = true, title::String = "default")
    l = length(state)
    outcomes = collect(0:l)
    #amplitudes (colors based on vector)
    if (amp)
        data = round.(abs.(state), digits = 3)
        bar_colors = [HSV(h, 1, v) for (h, _, v) in complex_to_hsv.(Complex.(state))]
        ylabel = "Amplitude"
#         if (title == "default")
#             title = "Outcome Amplitudes"
#         end
    #probabilities (all red)
    else
        data = round.(abs.(state).^2, digits = 3)
        bar_colors = [HSV(0,1,1) for a in state]
        ylabel = "Probability"
#         if (title == "default")
#             title = "Outcome Probabilities"
#         end
    end


    tufte_bar = Theme(
        default_color = colorant"green",
        background_color = colorant"white",
        bar_spacing = 2pt,
        grid_line_width = 0.25pt,
        minor_label_font = "Gill Sans",
        major_label_font = "Gill Sans",
        alphas = [0.7], key_position=:none
    )

    set_default_plot_size(max(10, l) * cm, 10 * cm)
    plot(x = [string(i) * " = " * string(i, base=2, pad=Int(log2(l))) for i in 0:(l - 1)],
         y = data, label=string.(data),
         Guide.xticks(ticks = outcomes, orientation = :vertical),
         Guide.yticks(ticks=[i * 0.1 for i = 0:10]),
         Coord.cartesian(ymin = 0.0, ymax = 1.1),
         Geom.bar, Geom.label(position=:above),
         Guide.xlabel("Outcomes"),
         Guide.ylabel(ylabel),
         Guide.title(title),
         color=0:l-1,
         Scale.color_discrete_manual(bar_colors...),
         tufte_bar)
end

function pixel_state(state::State, title::String = "Outcome Amplitudes")
    l = size(state)[1]
    outcomes = [0:l;]
    pixel_colors = [HSV(h, s, v) for (h, s, v) in complex_to_hsv.(Complex.(state))]

    tufte_bar = Theme(
        default_color=colorant"gray",
        background_color=colorant"white",
        bar_spacing=1pt,
        grid_line_width=0pt,
        minor_label_font="Gill Sans",
        major_label_font="Gill Sans",
        key_position=:none
    )

    set_default_plot_size(4cm, max(10, l)*cm)
    probabilities = round.(abs.(state).^2, digits=3)

    plot(y=[string(i)*"="*string(i, base=2, pad=Int(log2(l))) for i in (0:l-1)], x=[0.05 for _=0:l-1],
        Guide.ylabel(""), Guide.yticks(ticks=outcomes),
        Guide.xlabel(""), Guide.xticks(ticks=nothing),
        Coord.cartesian(xmin=0.0, xmax=0.1),
        Geom.bar(orientation=:horizontal),
        Guide.title(title),
        tufte_bar,
        color=0:l-1, Scale.color_discrete_manual(pixel_colors...))
end

import Cairo, Fontconfig

function save_state(state::State, title::String = "default", amplitude::Bool = true, bar::Bool = true)
    global counter += 1
    if (bar)
        draw(PNG(joinpath(folder * string(counter) * ".png"), max(10, length(state)) * cm, 10cm), bar_state(state, amplitude, title))
    else
        draw(PNG(joinpath(folder * string(counter) * ".png"), max(10, length(state)) * cm, 10cm), pixel_state(state, title))
    end
end

##################################### MATRICES ##########################################

function create_matrix_helper(trans::Transform, n::Int, cond::Function = f(x::Int)::Bool = true)

    G = Matrix{Complex}(I, n, n)

    target = Int(log2(n)) - trans.target - 1
    gate = trans.gate

    factor = Int(2 ^ (target + 1))
    shift =  Int(2 ^ (target))

    for prefix in 0:(n / factor - 1)
        for suffix in 0:(2^target - 1)
            m0 = Int(factor * (prefix) + suffix)
            if (cond(m0))
                # + 1 for julia indexing
                m0 += 1
                m1 = m0 + shift
                G[m0, m0] = gate.weights[1][1]
                G[m0, m1] = gate.weights[1][2]
                G[m1, m0] = gate.weights[2][1]
                G[m1, m1] = gate.weights[2][2]
            end
        end
    end

    return G

end

function create_matrix(trans::Transform, n::Int)
    mat = create_matrix_helper(trans, n)
    return mat
end

################################# INITIALIZATION ########################################

# initializes the state as an array of size 2^n
# index 1 has proportion 1, the rest have nothing (proportion 0)
function init_state(n::Int)::Array{Amplitude, 1}
    state::State = [0 for _ in 1:2 ^ n]
    state[1] = 1
    return state
end



